components/places/src/api/places_api.rs

```diff
@@ -82,9 +110,13 @@ pub struct PlacesApi {
     write_connection: Mutex<Option<PlacesDb>>,
     sync_state: Mutex<Option<SyncState>>,
     coop_tx_lock: Arc<Mutex<()>>,
+    // Used for get_sync_connection()
+    sync_connection: Mutex<Weak<Mutex<PlacesDb>>>,
```

Comment:
Doesn't look pretty to have a mutex in a mutex, but makes sense I'm assuming we need this to have interior mutability and keep it the `PlacesApi` `Sync`?

Reply:
Definitely want to second adding comments or refactoring here. I was initially confused but the explanation makes sense.

Reply:
BTW, one reason for the design is that I wanted to keep in place the current connection sharing behavior: We create a connection when syncing starts, and share it between the sync engines.  Once they are all done, we drop the connection.

I think we could lose the outer Mutex if we went for a different model, like always having one connection open, or opening the connection on the first sync then keeping it open.

Reply:
> BTW, one reason for the design is that I wanted to keep in place the current connection sharing behavior: We create a connection when syncing starts and share it between the sync engines. Once they are all done, we drop the connection.

I think it would be perfectly fine for 2 x Connections to be opened per sync, and also think it would be fine there was 1 lazy immortal one. We should avoid being wasteful, but I think it's small beer here?

Reply:
I think lazy immortal would be great, I can make that change.

Opening a new connection each time seems fine from a resource point of view, but I like having the Mutex to avoid any possibility of having 2 bookmarks/history engines updating the DB at the same time.

Reply:
Hmm, maybe it's not so easy.  I wanted to use `SyncLazy` to do this, but that's only available in nightly.  It wouldn't be hard to write a struct that implemented a similar API, but that doesn't reduce the complexity that much.  I think I'm just going to keep the current implementation, with the doc strings.